torch/distributed/_composable/fsdp/_fsdp_state.py

```diff
@@ -2,59 +2,74 @@
 
 import torch
 import torch.nn as nn
+from torch.autograd import Variable
+from torch.autograd.graph import register_multi_grad_hook
 
 from torch.distributed._composable_state import (
     _get_module_state,
     _insert_module_state,
     _State,
 )
 from torch.distributed.utils import _to_kwargs
+from torch.utils._pytree import tree_flatten
 from torch.utils.hooks import RemovableHandle
-from ._fsdp_collectives import AllGatherStateHolder
 from ._fsdp_common import TrainingState
 from ._fsdp_param import FSDPParam
+from ._fsdp_param_group import FSDPCommContext, FSDPParamGroup
 
-from ._fsdp_param_group import FSDPParamGroup
+
+class FSDPStateContext:
+    """This has state shared across FSDP states."""
+
+    def __init__(self):
+        # All FSDP states in the root state's module tree
+        self.all_states: List[FSDPState] = []
+        # Iteration's forward root runs the once-per-forward logic; this root
+        # may not be the overall root set by lazy initialization in cases where
+        # only a submodule runs forward
+        self.iter_forward_root: Optional[FSDPState] = None
```

Comment:
hmmm not fully get this, why this `iter_forward_root` could be different from the root after lazy init?